_posts/2021-09-06-connector-table-sql-api-part2.md

```diff
@@ -0,0 +1,518 @@
+
+public class ImapSource extends RichSourceFunction<RowData> {
+    @Override
+    public void run(SourceFunction.SourceContext<RowData> ctx) throws Exception {
+        // …
+
+        folder.addMessageCountListener(new MessageCountAdapter() {
+            @Override
+            public void messagesAdded(MessageCountEvent e) {
+                collectMessages(ctx, e.getMessages());
+            }
+        });
+
+        // …
+    }
+
+    private void collectMessages(SourceFunction.SourceContext<RowData> ctx, Message[] messages) {
+        for (Message message : messages) {
+            try {
+                ctx.collect(GenericRowData.of(StringData.fromString(message.getSubject())));
+            } catch (MessagingException ignored) {}
+        }
+    }
+}
+```
+
+Now build the project again and start up the SQL client:
+
+```sh
+$ cd testing/
+$ ./build_and_run.sh
+```
+
+This time, you will connect to a [GreenMail server](https://greenmail-mail-test.github.io/greenmail/) which is started as part of the [setup](https://github.com/Airblader/blog-imap/blob/master/testing/docker-compose.yaml):
+
+```sql
+CREATE TABLE T (
+    subject STRING
+) WITH (
+    'connector' = 'imap', 
+    'host' = 'greenmail',
+    'port' = '3143', 
+    'user' = 'alice', 
+    'password' = 'alice'
+);
+
+SELECT * FROM T;
+```
+
+The query above should now run continuously but no rows will be produced since it is a test server. You need to first send an email to the server. If you have [mailx](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/mailx.html) installed, you can do so by executing in your terminal:
+
+```sh
+$ echo "This is the email body" | mailx -Sv15-compat \
+        -s"Email Subject" \
+        -Smta="smtp://alice:alice@localhost:3025" \
+        alice@acme.org
+```
+
+The row “Email Subject” should now have appeared as a row in your output. Your source connector is working!
+
+However, since you are still hard-coding the schema produced by the source, defining the table with a different schema will produce errors. You want to be able to define which fields of an email interest you and then produce the data accordingly. To do this, you will use the list of column names from earlier and then look at it when you collect the emails. 
+
+```java
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.TimestampData;
+
+public class ImapSource extends RichSourceFunction<RowData> {
+
+    private void collectMessages(SourceFunction.SourceContext<RowData> ctx, Message[] messages) {
+        for (Message message : messages) {
+            try {
+                collectMessage(ctx, message);
+            } catch (MessagingException ignored) {}
+        }
+    }
+
+    private void collectMessage(SourceFunction.SourceContext<RowData> ctx, Message message)
+        throws MessagingException {
+        final RowData row = new GenericRowData(columnNames.size());
```

Comment:
Sorry, this was my mistake – this _needs_ to be changed as otherwise it's an error. 

```suggestion
        final GenericRowData row = new GenericRowData(columnNames.size());
```